src/base/dof_map_constraints.C

```diff
@@ -2165,6 +2166,65 @@ void DofMap::enforce_constraints_exactly (const System & system,
   v->close();
 }
 
+void DofMap::enforce_constraints_on_residual (const NonlinearImplicitSystem & system,
+                                              NumericVector<Number> * rhs,
+                                              NumericVector<Number> const * solution) const
+{
+  parallel_object_only();
+
+  if (!this->n_constrained_dofs())
+    return;
+
+  if (!rhs)
+    rhs = system.rhs;
+  if (!solution)
+    solution = system.solution.get();
+
+  libmesh_assert_equal_to (this, &(system.get_dof_map()));
+
+  for (const auto & pr : _dof_constraints)
+    {
+      dof_id_type constrained_dof = pr.first;
+      if (!this->local_index(constrained_dof))
+        continue;
+
+      const DofConstraintRow constraint_row = pr.second;
+
+      Number exact_value = 0;
+      for (const auto & j : constraint_row)
+        exact_value += j.second * (*solution)(j.first);
+      exact_value -= (*solution)(constrained_dof);
+
+      rhs->set(constrained_dof, exact_value);
+    }
+}
+
+void DofMap::enforce_constraints_on_jacobian (const NonlinearImplicitSystem & system,
+                                              SparseMatrix<Number> * jac) const
+{
+  parallel_object_only();
+
+  if (!this->n_constrained_dofs())
+    return;
+
+  if (!jac)
+    jac = system.matrix;
+
+  libmesh_assert_equal_to (this, &(system.get_dof_map()));
+
+  for (const auto & pr : _dof_constraints)
+    {
+      dof_id_type constrained_dof = pr.first;
+      if (!this->local_index(constrained_dof))
+        continue;
+
+      const DofConstraintRow constraint_row = pr.second;
+
+      for (const auto & j : constraint_row)
+        jac->set(constrained_dof, j.first, j.second);
+      jac->set(constrained_dof, constrained_dof, -1);
```

Comment:
This is going to force the equivalent of the asymmetric_constraint_rows option and it's going to prevent constrained problem Jacobian matrices from being SPD?  Is there any way around that?

Reply:
The Jacobian has to follow from the form of the residual statement, otherwise it will just be wrong. We'd need to think up our residual such that it naturally produces symmetric dependence on the dofs, e.g. if our residual function for the constrained dof depends on a couple of unconstrained dofs, then the residual of those unconstrained dofs must depend on the constrained dof...which would be a paradox...right?

Reply:
Well, the way we get a symmetric "jacobian" in most codes is to not let the solver worry about constrained dofs at all.  Instead of putting the constraint equation in the row for u_c, we put "u_c=0" in that row, and the Jacobian is just a 1 on the diagonal.  Then after a solve we enforce_constraints_exactly() to get the real value of each u_c correct.  The correct value isn't needed during a linear solve, because the constraint matrix pre and post multiplication zeros out the constrained dof's column, not just it's row.

This works well for most Newton-Krylov solves because even though we haven't tried to optimize enforce_constraints_exactly we only bother to call it after each linear solve step and the cost is trivial compared to that.  I'm not sure how well it would work for JFNK or for finite differenced Jacobians.

Reply:
> Well, the way we get a symmetric "jacobian" in most codes is to not let the solver worry about constrained dofs at all. Instead of putting the constraint equation in the row for u_c, we put "u_c=0" in that row, and the Jacobian is just a 1 on the diagonal. Then after a solve we enforce_constraints_exactly() to get the real value of each u_c correct. The correct value isn't needed during a linear solve, because the constraint matrix pre and post multiplication zeros out the constrained dof's column, not just it's row.
> 
> This works well for most Newton-Krylov solves

Yes I understand the idea and for many cases this works great. But the discussion in #1985 (hopefully now) illustrates that this doesn't work for JFNK or finite differenced jacobians.